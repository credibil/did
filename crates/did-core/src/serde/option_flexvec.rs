use std::fmt;
use std::marker::PhantomData;
use std::str::FromStr;

use serde::de;
use serde::de::{DeserializeOwned, Deserializer, SeqAccess, Visitor};
use serde::ser::{Serialize, SerializeSeq, Serializer};

pub fn serialize<T, S>(value: &Option<Vec<T>>, serializer: S) -> Result<S::Ok, S::Error>
where
    T: Serialize,
    S: Serializer,
{
    // serialize None as empty array
    if value.is_none() {
        return serializer.serialize_none();
    }

    let some_val = value.as_ref().expect("expected value but got none");

    let mut seq = serializer.serialize_seq(Some(some_val.len()))?;
    for e in some_val {
        seq.serialize_element(e)?;
    }
    seq.end()
}

pub fn deserialize<'de, T, D>(deserializer: D) -> Result<Option<Vec<T>>, D::Error>
where
    T: DeserializeOwned + FromStr,
    D: Deserializer<'de>,
{
    struct VisitorImpl<T>(PhantomData<fn() -> Vec<T>>);

    impl<'de, T> Visitor<'de> for VisitorImpl<T>
    where
        T: DeserializeOwned + FromStr,
    {
        type Value = Option<Vec<T>>;

        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
            formatter.write_str("Option<Vec<<T>>")
        }

        // deserialize array to Vec<T>
        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
        where
            A: SeqAccess<'de>,
        {
            // could be mixed array of strings and objects
            let mut deser: Vec<T> = Vec::new();
            while let Some(curr) = seq.next_element::<serde_json::Value>()? {
                match curr {
                    serde_json::Value::String(s) => {
                        let Ok(res) = T::from_str(&s) else {
                            return Err(de::Error::invalid_type(de::Unexpected::Str(&s), &self));
                        };
                        deser.push(res);
                    }
                    serde_json::Value::Object(o) => {
                        let Ok(res) = serde_json::from_value::<T>(serde_json::Value::Object(o))
                        else {
                            return Err(de::Error::invalid_type(de::Unexpected::Map, &self));
                        };
                        deser.push(res);
                    }
                    _ => {
                        return Err(de::Error::custom(
                            "invalid type: cannot deserialize array element",
                        ));
                    }
                }
            }
            Ok(Some(deser))
        }
    }

    deserializer.deserialize_any(VisitorImpl(PhantomData))
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use anyhow::anyhow;
    use serde::{Deserialize, Serialize};
    use serde_json::json;

    use crate::serde::option_flexvec;

    #[derive(Clone, Debug, Default, Deserialize, Serialize)]
    #[serde(default)]
    struct TestData {
        #[serde(with = "option_flexvec")]
        object_array: Option<Vec<Nested>>,

        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(with = "option_flexvec")]
        array: Option<Vec<String>>,

        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(with = "option_flexvec")]
        none: Option<Vec<String>>,
    }

    #[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
    struct Nested {
        n: String,
    }

    impl FromStr for Nested {
        type Err = anyhow::Error;

        fn from_str(_: &str) -> Result<Self, Self::Err> {
            Err(anyhow!("unimplemented"))
        }
    }

    #[test]
    fn option_flexvec() {
        let test_data = TestData {
            object_array: Some(vec![
                Nested {
                    n: "object1".to_string(),
                },
                Nested {
                    n: "object2".to_string(),
                },
            ]),
            array: Some(vec!["item1".to_string(), "item2".to_string()]),
            none: Default::default(),
        };

        // serialize
        let test_json = serde_json::to_value(&test_data).expect("failed to serialize");
        assert_eq!(
            *test_json.get("object_array").expect("expected value but got none"),
            json!([{"n": "object1"}, {"n": "object2"}])
        );
        assert_eq!(
            *test_json.get("array").expect("expected value but got none"),
            json!(["item1", "item2"]),
        );
        assert_eq!(test_json.get("none"), None);

        // deserialize
        let test_de: TestData = serde_json::from_value(test_json).expect("failed to deserialize");
        assert_eq!(test_de.object_array, test_data.object_array);
        assert_eq!(test_de.array, test_data.array);
        assert_eq!(test_de.none, test_data.none);
    }
}
