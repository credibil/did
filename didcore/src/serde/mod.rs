pub mod flexvec_or_single;
pub mod option_flexvec;
pub mod option_flexvec_or_single;
