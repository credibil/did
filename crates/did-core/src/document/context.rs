//! Context is a JSON-LD context that describes a DID Document schema.

use std::collections::HashMap;
use std::convert::Infallible;
use std::str::FromStr;

use serde::ser::SerializeMap;
use serde::{Deserialize, Serialize};

/// The default context for a DID document. <https://www.w3.org/TR/did-core/#context>
pub const DID_CONTEXT: &str = "https://www.w3.org/ns/did/v1";

/// Context uses JSON-LD, a JSON-based format used to serialize Linked Data. This section defines
/// the production and consumption rules for the JSON-LD representation.
///
/// The JSON-LD representation defines the following representation-specific entries:
///
/// @context
/// The JSON-LD Context is either a string or a list containing any combination of strings and/or
/// ordered maps.
#[derive(Clone, Debug, Default, Deserialize)]
pub struct Context {
    /// A single JSON-LD term.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    /// A map of JSON-LD terms.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url_map: Option<HashMap<String, String>>,
}

/// `FromStr` implementation for `Context`. Needed for deserializer.
impl FromStr for Context {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            url: Some(s.to_string()),
            url_map: None,
        })
    }
}

/// Serialize a context to a string or map. If only the `term` field is set, serialize to a string,
/// otherwise serialize to a map.
impl Serialize for Context {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match (&self.url, &self.url_map) {
            (Some(u), None) => serializer.serialize_str(u),
            (None, Some(map)) => {
                let mut m = serializer.serialize_map(Some(map.len()))?;
                for (k, v) in map {
                    m.serialize_entry(k, v)?;
                }
                m.end()
            }
            _ => Err(serde::ser::Error::custom("Context must be a string or map")),
        }
    }
}

pub(crate) mod context_serialization {
    use std::collections::HashMap;
    use std::fmt;
    use std::marker::PhantomData;
    use std::str::FromStr;

    use serde::de::{self, SeqAccess, Visitor};
    use serde::ser::{SerializeSeq, Serializer};

    use super::Context;

    pub(crate) fn serialize<S>(value: &[Context], serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        if value.len() == 1 && value[0].url.is_some() && value[0].url_map.is_none() {
            serializer.serialize_some(&value[0])
        } else {
            let mut seq = serializer.serialize_seq(Some(value.len()))?;
            for e in value {
                seq.serialize_element(e)?;
            }
            seq.end()
        }
    }

    pub(crate) fn deserialize<'de, D>(deserializer: D) -> Result<Vec<Context>, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        struct StringOrMap<T>(PhantomData<fn() -> Vec<T>>);

        impl<'de> Visitor<'de> for StringOrMap<Vec<Context>> {
            type Value = Vec<Context>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("Context as a single string or a set of strings and/or maps")
            }

            // If a string is found, deserialize it to a single entry Vec<Context> using Context::from_str
            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                match Context::from_str(value) {
                    Ok(res) => Ok(vec![res]),
                    Err(_) => Err(de::Error::invalid_type(de::Unexpected::Str(value), &self)),
                }
            }

            // Deserialize a sequence to Vec<Context>
            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                // could be mixed array of strings and objects
                let mut deser: Vec<Context> = Vec::new();
                while let Some(curr) = seq.next_element::<serde_json::Value>()? {
                    match curr {
                        serde_json::Value::String(s) => {
                            let Ok(res) = Context::from_str(&s) else {
                                return Err(de::Error::invalid_type(
                                    de::Unexpected::Str(&s),
                                    &self,
                                ));
                            };
                            deser.push(res);
                        }
                        serde_json::Value::Object(o) => {
                            let Ok(res) = serde_json::from_value::<HashMap<String, String>>(
                                serde_json::Value::Object(o),
                            ) else {
                                return Err(de::Error::invalid_type(de::Unexpected::Map, &self));
                            };
                            let ctx = Context {
                                url: None,
                                url_map: Some(res),
                            };
                            deser.push(ctx);
                        }
                        _ => {
                            return Err(de::Error::custom(
                                "invalid type: cannot deserialize array element",
                            ));
                        }
                    }
                }
                Ok(deser)
            }
        }

        deserializer.deserialize_any(StringOrMap(PhantomData))
    }
}

#[cfg(test)]
mod tests {

    use olpc_cjson::CanonicalFormatter;

    use super::*;

    #[derive(Serialize, Deserialize)]
    struct Doc {
        #[serde(with = "context_serialization", rename = "@context")]
        pub context: Vec<Context>,
    }

    #[test]
    fn deserialize_single() {
        let input = r#"{
            "@context": "https://www.w3.org/ns/did/v1"
        }"#;

        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_strings() {
        let input = r#"{
            "@context": [
                "https://www.w3.org/ns/did/v1",
                "https://w3id.org/security/suites/jws-2020/v1"
            ]
        }"#;

        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_maps() {
        let input = r#"{
            "@context": [
                { "@base": "https://example.com/keys/" }
            ]
        }"#;

        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_mix() {
        let input = r#"{
            "@context": [
                "https://www.w3.org/ns/did/v1",
                "https://w3id.org/security/suites/jws-2020/v1",
                { "@base": "https://example.com/keys/" }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        })
    }

    #[test]
    fn serialize_single() {
        let doc = Doc {
            context: vec![Context {
                url: Some("https://www.w3.org/ns/did/v1".to_string()),
                url_map: None,
            }],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(json, r#"{"@context":"https://www.w3.org/ns/did/v1"}"#);
    }

    #[test]
    fn serialize_maps() {
        let doc = Doc {
            context: vec![Context {
                url: None,
                url_map: Some(
                    vec![("@base".to_string(), "https://example.com/keys/".to_string())]
                        .into_iter()
                        .collect(),
                ),
            }],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(json, r#"{"@context":[{"@base":"https://example.com/keys/"}]}"#);
    }

    #[test]
    fn serialize_mix() {
        let doc = Doc {
            context: vec![
                Context {
                    url: Some("https://www.w3.org/ns/did/v1".to_string()),
                    url_map: None,
                },
                Context {
                    url: Some("https://w3id.org/security/suites/jws-2020/v1".to_string()),
                    url_map: None,
                },
                Context {
                    url: None,
                    url_map: Some(
                        vec![
                            ("@base".to_string(), "https://example.com/keys/".to_string()),
                            ("other".to_string(), "value".to_string()),
                        ]
                        .into_iter()
                        .collect(),
                    ),
                },
                Context {
                    url: None,
                    url_map: Some(HashMap::from([(
                        "api".to_string(),
                        "https://example.com/api/".to_string(),
                    )])),
                },
            ],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(
            json,
            r#"{"@context":["https://www.w3.org/ns/did/v1","https://w3id.org/security/suites/jws-2020/v1",{"@base":"https://example.com/keys/","other":"value"},{"api":"https://example.com/api/"}]}"#
        );
    }
}
