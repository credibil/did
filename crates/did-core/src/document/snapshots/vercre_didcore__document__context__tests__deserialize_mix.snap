---
source: did-core/src/document/context.rs
assertion_line: 213
expression: doc
---
"@context":
  - "https://www.w3.org/ns/did/v1"
  - "https://w3id.org/security/suites/jws-2020/v1"
  - "@base": "https://example.com/keys/"

